From: Harald Freudenberger <freude@linux.vnet.ibm.com>
Subject: crypto: kernel oops at insmod of the z90crypt device driver
Patch-mainline: not yet
Git-commit: -
References: bnc#909088, LTC#119591

Description:  crypto: kernel oops at insmod of the z90crypt device driver
Symptom:      Kernel oops caused by invalid parameter at TAPQ instruction
Problem:      On older systems where the QCI instruction is not available
              all possible domains are probed via TAPQ instruction. The
              range for the probe has been extended with the > 16 domain
              support now leading to a possible specification exception
              when this instruction is called for probing higher values
              within the new range. This may happen during insmod and/or
              ap bus reset only on machines without a QCI instruction (z10,
              z196, z114), zEC12 and newer systems are not affected.
Solution:     Fix the domain checking function to limit the allowed range
              if no QCI info is available.
Reproduction: insmod z90crypt on z196 or z10 with at least one crypto card
              (eg. CEX4) installed and configured for the LPAR or VM.

Signed-off-by: Harald Freudenberger <freude@linux.vnet.ibm.com>
Acked-by: John Jolly <jjolly@suse.de>
---
 drivers/s390/crypto/ap_bus.c |   10 +++++++---
 1 file changed, 7 insertions(+), 3 deletions(-)

--- a/drivers/s390/crypto/ap_bus.c
+++ b/drivers/s390/crypto/ap_bus.c
@@ -1157,9 +1157,13 @@ static inline int ap_test_config_card_id
  */
 static inline int ap_test_config_domain(unsigned int domain)
 {
-	if (!ap_configuration)
-		return 1;
-	return ap_test_config(ap_configuration->aqm, domain);
+	if (!ap_configuration)    /* QCI not supported */
+		if (domain < 16)
+			return 1; /* then domains 0...15 are configured */
+		else
+			return 0;
+	else
+		return ap_test_config(ap_configuration->aqm, domain);
 }
 
 /**
